src/frontend/src/instance.rs

```diff
@@ -564,15 +580,26 @@ impl SqlQueryHandler for Instance {
         query: &str,
         query_ctx: QueryContextRef,
     ) -> Vec<server_error::Result<Output>> {
-        match parse_stmt(query)
+        let query_interceptor = self.plugins.get::<SqlQueryInterceptorRef>();
+        let query = match query_interceptor.pre_parsing(query, query_ctx.clone()) {
+            Ok(q) => q,
+            Err(e) => return vec![Err(e)],
+        };
+
+        match parse_stmt(query.as_ref())
             .map_err(BoxedError::new)
             .context(server_error::ExecuteQuerySnafu { query })
+            .and_then(|stmts| query_interceptor.post_parsing(stmts, query_ctx.clone()))
         {
             Ok(stmts) => {
                 let mut results = Vec::with_capacity(stmts.len());
                 for stmt in stmts {
                     match self.query_statement(stmt, query_ctx.clone()).await {
```

Comment:
Better invoke `pre_execute` here, even though `pre_execute` does nothing now. To pair with `post_execute`.

Reply:
I'm going to feed `pre_execute` with `LogicalPlan` so later we can evaluate the cost of a query in the interceptor. For now I just call `pre_execute` with `LogicialPlan` as `None` but we need to provide that information later.